cpp/src/io/utilities/block_utils.cuh

```diff
@@ -39,11 +39,8 @@ inline __device__ uint32_t ballot(int pred) { return __ballot_sync(~0, pred); }
 template <typename T>
 inline __device__ void nanosleep(T d)
 {
-#if (__CUDA_ARCH__ >= 700)
-  __nanosleep(d);
-#else
+  // TODO: nanosleep temporarily disabled 
   clock();
-#endif
```

Comment:
I don't get it. Why the function `nanosleep()` returns the result of `clock()`? As in C++ reference, `clock()` "returns the processor time consumed by the program", nothing related to sleeping a thread.

Reply:
This is to workaround a driver bug that affects `nanosleep` performance.

Reply:
This is also in device code, not host code...

Reply:
Agree about the naming. Open for suggestions :-)
`nanosleep()` is currently only used inside decompression as a means of yielding (originally through `__nanosleep()`). We have had `clock()` as an alternative for SM <7.0 simply to avoid having it optimized away.

Reply:
Got it---"alternative for SM <7.0 simply to avoid having it optimized away". If possible, please add a comment line to it so in the next time other people like me can understand without asking 😄

Reply:
> We have had clock() as an alternative for SM <7.0 simply to avoid having it optimized away.

I don't understand this part. This is just busy polling, but with extra steps. If you wanted to emulate the "sleep" part of nanosleep as in "don't do anything for X clocks" with `clock()`, you could do:
```
auto start = clock();
while(start < (start = clock()) + X){};
```

Reply:
Thanks, @jrhemstad. Get your point. My intention for this workaround was to expose as little risk as possible of introducing a perf regression (or breaking something), so I went with what we already did for pre-Volta.

I'm also fine to completely drop the `clock()` detour. Longer term, I think it makes sense to get rid of the polling and task parallel approach altogether.

Reply:
@jrhemstad I agree about adding waiting logic to the wrapper `nanosleep()` function. Currently it works because
everywhere nanosleep is used, it is also waiting on its own. https://github.com/rapidsai/cudf/blob/1771a8fc18ac647c9172209fa3b39b40437ac85c/cpp/src/io/comp/gpuinflate.cu#L665 https://github.com/rapidsai/cudf/blob/2780a8cea7743002d62b72c804ef693c8807d586/cpp/src/io/comp/unsnap.cu#L493

But this also means that any new logic that strictly relies on the wait to happen in `nanosleep()` itself will break for pascal.

Reply:
Agreed with Jake to completely get rid of nanosleep and its usage.